src/main/python/docs/source/guide/python_end_to_end_tut.rst

```diff
@@ -0,0 +1,561 @@
+     ,{ "name": "capital-loss", "method": "global_mean" }
+     ,{ "name": "hours-per-week", "method": "global_mean" }
+     ,{ "name": "native-country"        , "method": "global_mode" }
+    ],
+    "bin": [ { "name": "age"  , "method": "equi-width", "numbins": 3 }],
+    "dummycode": ["age", "workclass", "education", "marital-status", "occupation", "relationship", "race", "sex", "native-country"],
+    "recode": ["income"]
+    }
+
+Our dataset has missing values. An easy way to deal with that circumstance is to use the "impute" option that SystemDS supports.
+We simply pass a list that holds all the relations between column names and the method of interpolation. A more specific example  is the "education" column.
+In the dataset certain entries have missing values for this column. As this is a string feature,
+we can simply define the method as "global_mode" and replace every missing value with the global mode inside this column. It is important to note that
+we first had to define the values of the missing strings in our selected dataset using the .mtd files (naStrings": [ "?", "" ]).
+
+With the "bin" keyword we can discretize continuous values into a small number of bins. Here the column with age values
+is discretized into three age intervals. The only method that is currently supported is equi-width binning.
+
+The column-level data transformation "dummycode" allows us to one-hot-encode a column.
+In our example we first bin the "age" column into 3 different bins. This means that we now have one column where one entry can belong to one of 3 age groups. After using
+"dummycode", we transform this one column into 3 different columns, one for each bin.
+
+At last, we make use of the "recode" transformation for categorical columns, it maps all distinct categories in
+the column into consecutive numbers, starting from 1. In our example we recode the "income" column, which
+transforms it from "<=$50K" and ">$50K" to "1" and "2" respectively.
+
+Another good resource for further ways of processing is: https://apache.github.io/systemds/site/dml-language-reference.html
+
+There we provide different examples for defining jspec's and what functionality is currently supported.
+
+After defining the .jspec file we can read it by passing the filepath, data_type and value_type using the following command:
+
+.. code-block:: python
+
+    dataset_jspec = "adult/jspec.json"
+    jspec = sds.read(dataset_jspec, data_type="scalar", value_type="string")
+
+Finally, we need to define a custom dml file to split the features from the labels and replace certain values, which we will need later.
+We will call this file "preprocess.dml":
+
+.. code-block::
+
+    get_X = function(matrix[double] X, int start, int stop)
+    return (matrix[double] returnVal) {
+    returnVal = X[start:stop,1:ncol(X)-1]
+    }
+    get_Y = function(matrix[double] X, int start, int stop)
+    return (matrix[double] returnVal) {
+    returnVal = X[start:stop,ncol(X):ncol(X)]
```

Comment:
indent

Reply:
ah i see, you make this function to enable slicing columns out.

I will add a slice function later, than remove the need of using this.